Implement basic support for the field reflection API

cyang1 opened the pull request:
Most of the work here involved adding blanket-implemented erased typings of most of the types in the crate, so that they could be referenced as trait objects from the new `fields` method.

This involved splitting up the `Message` trait into a separate `Serialize` and `Deserialize`, as I didn't have a use-case for erased deserialization.

I'm not entirely sure all the type-bound changes here are reasonable or the best way to do things.

nipunn1313 commented:
Gotta commit the updated generated code in the tests - so that the tests pass (and so we can see the generated code in the code review, which makes it easier to see what's going on).

nipunn1313 commented on pb-jelly-gen/codegen/codegen.py:
```diff
@@ -1012,7 +1011,29 @@ def gen_msg(
             )
 
-        with block(self, "impl ::pb_jelly::Message for " + name):
+        with block(self, "impl ::pb_jelly::Serialize for " + name):
```
`Serialize` and `Deserialize` are special words in `serde` and I think it'll be confusing to reuse the same terminology here - since it's not really referring to the same thing (and it's similarish?)

cyang1 replied:
I feel like they're common enough verbs that it won't be too confusing to see them used in another crate. These also happen to be the same verbs used in the (majority) of the public methods in each of these traits. Any other verbs I can think of would just be more confusing IMO. Were you able to come up with good alternative names?

nipunn1313 commented:
What is the field reflection API and the purpose of adding support here? This seems like a lot of code to give some kind of Box Dyn? There's a lot of type bound changes that seem irrelevant? I think some more context to the motive here would be helpful.

cyang1 commented:
The field reflection API is primarily described by the `FieldDescriptorProto` in https://github.com/protocolbuffers/protobuf/blob/master/src/google/protobuf/descriptor.proto, which is part of the `DescriptorProto` https://github.com/protocolbuffers/protobuf/blob/7ccf4d8f67878a6ceb2184df279478cb3314372b/src/google/protobuf/descriptor.proto#L97.

Something I've added here (which results in most of the added logic in this diff) is a borrow of each individual field as part of the `FieldDescriptorProto`. This isn't necessary in more dynamic languages like Python, where you're able to retrieve the value of the field from just the field name, but it is something we have to explicitly include in Rust. However, because we're returning a collection, the type we give to the field value must be uniform. I originally wanted to do something like `&'a dyn Message`, but the existing `Message` trait is not object-safe for more than one reason (methods with generic parameters, and supertrait requirements), so I added blanket-implemented erased typings, which is what you see in `erased.rs`. This is akin to something like https://github.com/dtolnay/erased-serde, which solves the same issue for the `serde` types.

My primary use-case for this change is to be able to serialize each field of a proto individually, agnostic to the type of the proto specifically.

After some testing, I need to think more about what to do about `oneof` and `repeated` fields. Might change the structure of this a little bit.

nipunn1313 commented:
Is there a way to solve this problem with much less complexity?
Do you need to have a collection of `&dyn Message` (or similar)?

Where do you expect to be using such a collection? If you're creating a `.fields()` method and immediately iterating through it an serializing, perhaps it can be done in a way where you don't need an erased type collection.

cyang1 commented:
I believe no matter how this is done, it's always going to need a common trait that can be serialized, and (I believe) the minimal amount of work to do that is everything in `erased.rs`.

Consider this as in "planning changes" for now though, I want to think about the API a bit more.

cyang1 commented:
Reorganized a bit of stuff. Decided to add the relevant `FieldDescriptor` information into `MessageDescriptor`, and added a new `Reflection` trait that implements borrowing the field. If the general shape of this looks good, I can add some comments and probably a test or two.

cyang1 commented:
Now simpler! Also added a test + comments.

cyang1 commented:
By the way, I'm not sure how to make the tests pass here. Locally, I was testing this by changing all the `Cargo.toml`s under `pb-test` to point to `pb-jelly` by path rather than version so that it picked up my changes. Is that something we should do in general?

goffrie commented:
> By the way, I'm not sure how to make the tests pass here. Locally, I was testing this by changing all the `Cargo.toml`s under `pb-test` to point to `pb-jelly` by path rather than version so that it picked up my changes. Is that something we should do in general?

We should add a `[patch]` to `pb-test/Cargo.toml`. I'll make a separate PR for this.

goffrie commented:
#128

ddeville commented:
Seems like the example job pulls version 0.0.9 from crates.io so obviously that will fail until we update the crate. That seems sub-optimal 🤷🏼‍♂️

nipunn1313 commented:
cool!
Is there any risk that this might break any existing APIs? It seems like it's purely an addition, so hopefully no - but if so, we should look at those parts carefully and consider feature gating it.
Nice.

cyang1 commented:
Hmm, it looks like `examples/src/zero_copy/main.rs` is broken on `main` (along with examples not being patched correctly). I can fix it here, although I don't know that that's the best way to organize this.

goffrie commented:
> Hmm, it looks like `examples/src/zero_copy/main.rs` is broken on `main` (along with examples not being patched correctly). I can fix it here, although I don't know that that's the best way to organize this.

#130
